### PR TITLE
Add optional JSON-Schema validation, metadata arg, and pytest

### LINKED ISSUE #3: Initial todos
- [ ] Remove the builtin method param validation (looks like kbase copy-and-paste versions of this library remove it and don't use it)
- [ ] Add optional metadata param for `call()` that gets passed down to the actual method. Useful for things like an auth token from a header.
- [ ] Add optional jsonschema validation for methods

### PR DESCRIPTION
**Pull request by jayrbolton**
This addresses all the todos from https://github.com/kbaseIncubator/jsonrpcbase/issues/3

* Add a metadata argument to the method call, useful for things such as an auth token, which might come from an HTTP header
* Add optional JSON-Schema method validation and remove the old validation logic
* Convert from nose to pytest

### CODE REVIEW

File: jsonrpcbase/main.py [hunk 1]
```diff
@@ -54,8 +54,7 @@ def send_message(msg):
 import json
 import logging
 import types
-
-import six
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:58]: We don't need python 2 compatibility

File: jsonrpcbase/main.py [hunk 2]
```diff
@@ -69,8 +68,9 @@ class JSONRPCService(object):
 
     def __init__(self):
         self.method_data = {}
+        self.json_schemas = {}
 
-    def add(self, f, name=None, types=None, required=None):
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:73]: `types` and `required` arguments replaced with a single jsonschema arg

File: jsonrpcbase/main.py [hunk 3]
```diff
@@ -130,26 +123,17 @@ def call_py(self, jsondata):
...
         request = self._get_default_vals()
-
         try:
             if isinstance(rdata, dict) and rdata:
                 # It's a single request.
                 self._fill_request(request, rdata)
-                respond = self._handle_request(request)
-
-                # Don't respond to notifications
-                if respond is None:
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:144]: This block does nothing

File: poetry.lock [hunk 4]
```diff
@@ -1,3 +1,35 @@
+[[package]]
```
Comment (jayrbolton) [inline poetry.lock:1]: See pyproject.toml

File: test/test_jsonrpcbase.py [hunk 5]
```diff
@@ -1,40 +1,37 @@
...
 import jsonrpcbase
-from nose.tools import assert_equal, assert_not_equal
-import six
 
 s = jsonrpcbase.JSONRPCService()
 
 
-def subtract(a, b):
-    return a - b
+def subtract(params, meta):
```
Comment (jayrbolton) [inline test/test_jsonrpcbase.py:11]: All methods now have these uniform parameters, where `params` is the deserialized json data

File: jsonrpcbase/main.py [hunk 6]
```diff
@@ -53,11 +55,9 @@ def send_message(msg):
 """
 import json
 import logging
-import types
-
-import six
+import jsonschema
 
-DEFAULT_JSONRPC = '2.0'
+DEFAULT_JSONRPC = (2, 0)
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:60]: Before they were using an integer of the version like `20`, but I think a tuple makes more sense (eg `1.11` and `11.1` would be ambiguous as an integer). Plus, tuples are comparable in python, so you can do `(1,1) < (2,0)`

File: jsonrpcbase/main.py [hunk 7]
```diff
@@ -165,122 +150,92 @@ def call_py(self, jsondata):
...
+        }
+        return resp
 
     def _get_err(self, e, id=None, jsonrpc=DEFAULT_JSONRPC):
         """
         Returns jsonrpc error message.
         """
         # Do not respond to notifications when the request is valid.
-        if not id and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
+        if id is None and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:206]: `not id` will be true if `id == 0`, which should be a valid id

File: jsonrpcbase/main.py [hunk 8]
```diff
@@ -165,122 +150,92 @@ def call_py(self, jsondata):
...
+        Fills version information to the respond from the internal tuple version.
         """
-        if iver == 20:
+        if ver == (2, 0):
             respond['jsonrpc'] = '2.0'
-        if iver == 11:
+        elif ver == (1, 1):
             respond['version'] = '1.1'
 
-    def _vargs(self, f):
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:231]: This validation stuff is removed and replaced with the jsonschema

File: jsonrpcbase/main.py [hunk 9]
```diff
@@ -329,133 +284,69 @@ def _get_params(self, rdata):
...
-                             or params[key] is None):
-                    raise InvalidParamsError('arg "%s" is the wrong type' % key)
+        ver = '.'.join(str(i) for i in DEFAULT_JSONRPC)
+        return {"jsonrpc": ver, "id": None}
 
 
 class JSONRPCError(Exception):
     """
     JSONRPCError class based on the JSON-RPC 2.0 specs.
-
```
Comment (jayrbolton) [inline jsonrpcbase/main.py:436]: Everything removed under this class is unused code

File: jsonrpcbase/__init__.py [hunk 10]
```diff
@@ -1,3 +1,3 @@
-from jsonrpcbase.main import JSONRPCService, DEFAULT_JSONRPC
+from jsonrpcbase.main import JSONRPCService
 
-__all__ = ['JSONRPCService', 'DEFAULT_JSONRPC']
```
Comment (jayrbolton) [inline jsonrpcbase/__init__.py:3]: DEFAULT_JSONRPC was only needed in the tests, which seemed silly, so I unexported it

File: jsonrpcbase/main.py [hunk 11]
```diff
@@ -165,122 +150,93 @@ def call_py(self, jsondata):
...
-                respond['error'] = e.dumps()['message']
-            else:
-                self._fill_ver(jsonrpc, respond)
-                respond['error'] = e.dumps()
-        else:
-            respond['jsonrpc'] = jsonrpc
-            respond['error'] = e.dumps()
-
+        self._fill_ver(jsonrpc, respond)
+        respond['error'] = e.dumps()
```
Comment (xpe) [inline jsonrpcbase/main.py:210]: Discussed. JSON RPC 1.1 and 2.0 have the same error style. (1.0 support removed)

File: jsonrpcbase/main.py (hunk 8 above)
Comment (xpe) [reply jsonrpcbase/main.py:231]: Nice to get rid of the homegrown stuff and replace with a nicer way!

File: jsonrpcbase/main.py [hunk 12]
```diff
@@ -289,14 +245,11 @@ def _get_id(self, rdata):
         InvalidRequestError will be raised if the id value has invalid type.
         """
         if 'id' in rdata:
-            if isinstance(rdata['id'], six.string_types) or \
-                    isinstance(rdata['id'], six.integer_types) or \
-                    isinstance(rdata['id'], float) or \
-                    rdata['id'] is None:
+            if type(rdata['id']) in (str, int, float):
```
Comment (xpe) [inline jsonrpcbase/main.py:248]: From [JSON RPC 2.0 spec][spec]:

> id
>
> An identifier established by the Client that MUST contain a String, Number, or NULL value if included. If it is not included it is assumed to be a notification. The value SHOULD normally not be Null [1] and Numbers SHOULD NOT contain fractional parts [2]

Let's do a follow-up to fix this, most likely with JSON schema.

[spec]: https://www.jsonrpc.org/specification#request_object

File: poetry.lock (hunk 4 above)
Comment (xpe) [reply poetry.lock:1]: Since it is generated by poetry according to pyproject.toml.

**Review by xpe — CHANGES_REQUESTED** [review]
Requested changes for this PR:

- [x] create follow-up issues as needed (see comments below)
- [x] be strict in requiring the JSON RPC version
- [ ] update docstrings and README
- [ ] ... with example of `metadata` and rationale for why it exists
- [x] add test case(s) for `metadata`
- [x] (later, so the action item is to create a ticket) remove JSON RPC 1.1 support (we'll elaborate on rationale later)
- [x] (later, so the action item is to create a ticket) don't allow floating-point id's. Jay likes the idea of using JSON Schema for this validation.

Otherwise, LGTM!

**Comment by jayrbolton** [comment]
Merging this and will address all the issues we discussed in followup PRs

**Pull request merged**